pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the finite-difference gradient
+of a quantum tape.
+"""
+# pylint: disable=protected-access,too-many-arguments
+import numpy as np
+from scipy.special import factorial
+
+import pennylane as qml
+
+
+def finite_diff_stencil(n, order, form):
+    """Generate the finite difference stencil (shift and coefficients)
+    for various derivatives, accuracy, and form.
+
+    Args:
+        n (int): Positive integer specifying the derivative. ``n=1``
+            corresponds to the first derivative.
+        order (int): Positive integer referring to the accuracy of the
+            returned stencil. E.g., ``order=1`` corresponds to a stencil
+            that returns a first-order approximation to the derivative.
+        form (str): one of ``"forward"``, ``"center"``, or ``"backward"``.
+
+    Returns:
+        array[float]: a ``(2, N)`` array. The first row corresponds to the
+        coefficients, and the second row corresponds to the shifts.
+
+    **Example**
+
+    >>> finite_diff_stencil(n=1, order=1, form="forward")
+    array([[-1.,  1.],
+           [ 0.,  1.]])
+    >>> finite_diff_stencil(n=1, order=2, form="center")
+    array([[-0.5,  0.5],
+           [-1. ,  1. ]])
+    >>> finite_diff_stencil(n=2, order=2, form="center")
+    array([[-2.,  1.,  1.],
+           [ 0., -1.,  1.]])
+    """
+    if n < 1 or not isinstance(n, int):
+        raise ValueError("Derivative order n must be a positive integer.")
+
+    if order < 1 or not isinstance(order, int):
+        raise ValueError("Accuracy order must be a positive integer.")

```

Comment:
"Accuracy" suggests to me some kind of small float tolerance rather than an integer; maybe "approximation order" or something? (same comment re. the argument description in the docstring)

Reply:
Done!